enhancements/kube-apiserver/generic-delete-protection.md

```diff
@@ -0,0 +1,233 @@
+}
+
+type FinalizerCriticalServiceCriteria struct{
+    GroupResource
+    FinalizerName string
+}
+
+type SpecificResourceCriticalServiceCriteria struct{
+    GroupResource
+    Namespace string
+    Name string
+}
+```
+
+### User Stories 
+
+#### Deployment provides finalizer processing for CRD
+```yaml
+kind: CriticalService
+spec:
+  provider:
+    group: apps
+    resource: deployments
+    namespace: finalizer-namespace
+    name: finalizer-deployment
+  criteria:
+  - type: Finalizer
+    finalizer:
+      group: my.crd.group
+      resource: coolresources
+      finalizerName: my.crd.group/super-important
+  - type: SpecificResource
+    specificResource:
+      group: apiextensions.k8s.io
+      resource: CustomResourceDefinition
+      name: coolresources.my.crd.group
+```
+
+
+This would allow three separate ManifestWorks to all be deleted at the same time and avoid conflicting with each other
+as the deletes are finalized on individual resources.
+```yaml
+kind: ManifestWork
+metadata:
+  name: finalizer
+spec:
+  criticalservices/for-finalizer-deployment
+  namespace/finalizer-namespace
+  deployment.apps/finalizer-deployment
+---
+kind: ManifestWork
+metadata:
+  name: crd
+spec:
+  crd/coolresources.my.crd.group
+---
+kind: ManifestWork
+metadata:
+  name: cr
+spec:
+  coolresources.my.crd.group/some-instance
+```
+
+It would also allow them to be combined into a single ManifestWork.
+```yaml
+kind: ManifestWork
+metadata:
+  name: all
+spec:
+  criticalservices/for-finalizer-deployment
+  namespace/finalizer-namespace
+  deployment.apps/finalizer-deployment
+  crd/coolresources.my.crd.group
+  coolresources.my.crd.group/some-instance
+```
+
+This construct also means that namespaces in a management cluster can be be deleted when managed clusters are removed
+because deletion can happen in any order. 
+
+
+When a bulk delete happens, the effective order will be
+1. crd/coolresources.my.crd.group is deleted, but waits to be finalized
+2. coolresources.my.crd.group/some-instance is deleted, but waits to be finalized
+3. coolresources.my.crd.group/some-instance is finalized
+4. crd/coolresources.my.crd.group is finalized
+5. deployment.apps/finalizer-deployment is deleted, but waits to be finalized
+6. namespace/finalizer-namespace is deleted, but waits to be finalized
+7. deployment.apps/finalizer-deployment is finalized
+8. namespace/finalizer-namespace is finalized
+9. criticalservices/for-finalizer-deployment is deleted
+
+This is because
+1. CRDs are not finalized until all the CR instances are removed
+2. deployment.apps/finalizer-deployment and namespace/finalizer-namespace cannot be deleted until the finalizer is
+   removed from all coolresources.my.crd.group and the crd/coolresources.my.crd.group is finalized.
+3. criticalservices/for-finalizer-deployment cannot be deleted until deployment.apps/finalizer-deployment is finalized
+
+This is enforced without client deletion coordination.
+
+#### Story 2
+
```

Comment:
Another story is that the provider and  criteria  could be the same kind of resources.  Taking manifestwork as an example. We could have a manifestwork defining the deployment of agent that operats manifetwork apis, for the purpose of agent upgrade in the future. It means there will be a `critical` manifestwork and other manifestworks, and the deletion of `critical` manifestwork should be after all the other manifestworks. I think CriticalService API should be able to handle such case also.